### PR TITLE
GLT-2202: Create workflow UI, change workflow database implementation

### PR DESCRIPTION
**Pull request by ndrein-oicr**
https://jira.oicr.on.ca/browse/GLT-2202

Apologies for this monolith pull request...the DAO/service layer changes were necessary to power the UI.

### CODE REVIEW

File: miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/WorkflowController.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+ * Responsible for serving workflow-related JSPs
+ */
+@RequestMapping("/workflow")
+@Controller
+public class WorkflowController {
+  @Autowired
+  WorkflowManager workflowManager;
+
+  @RequestMapping("/new/{workflowName}")
+  public ModelAndView createWorkflow(@PathVariable String workflowName, ModelMap model) throws IOException {
```
Comment (ndrein-oicr) [inline miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/WorkflowController.java]: These 2 endpoints are now accessible, but no links to these pages exist, by intention.

File: miso-web/src/main/webapp/scripts/workflow_display.js [hunk 2]
```diff
@@ -0,0 +1,167 @@
+WorkflowDisplay = (function() {
```
Comment (sqwci) [inline miso-web/src/main/webapp/scripts/workflow_display.js]: BLOCKER Add the "let", "const" or "var" keyword to this declaration of "WorkflowDisplay" to make it explicit. [rule](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS2703)

File: core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/workflow/impl/IntegerProgressStep.java [hunk 3]
```diff
@@ -1,8 +1,14 @@
 package uk.ac.bbsrc.tgac.miso.core.data.workflow.impl;
 
+import javax.persistence.Column;
```
Comment (sqwci) [inline core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/workflow/impl/IntegerProgressStep.java]: MINOR Remove this unused import 'javax.persistence.Column'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

File: core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/workflow/Progress.java [hunk 4]
```diff
@@ -10,11 +8,14 @@
 import com.eaglegenomics.simlims.core.User;
 
 import uk.ac.bbsrc.tgac.miso.core.data.Identifiable;
+import uk.ac.bbsrc.tgac.miso.core.data.workflow.Workflow.WorkflowName;
 
 /**
  * Models an in-progress workflow
  */
 public interface Progress extends Identifiable, Serializable {
```
Comment (sqwci) [inline core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/workflow/Progress.java]: CRITICAL Move constants to a class or enum. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1214)

**Comment by sqwci** [comment]
SonarQube analysis reported 7 issues
* ![BLOCKER][BLOCKER] 1 blocker
* ![CRITICAL][CRITICAL] 3 critical
* ![MINOR][MINOR] 3 minor

Watch the comments in this conversation to review them.

#### 4 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [LoadSequencerWorkflow.java#L43](https://github.com/TGAC/miso-lims/blob/91d7ed1a5a0bbcc3e7e45c662e9fba927873c0c2/core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/workflow/impl/LoadSequencerWorkflow.java#L43): Define a constant instead of duplicating this literal "Not implemented yet" 4 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [ProgressImpl.java#L136](https://github.com/TGAC/miso-lims/blob/91d7ed1a5a0bbcc3e7e45c662e9fba927873c0c2/core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/workflow/impl/ProgressImpl.java#L136): Refactor this method to reduce its Cognitive Complexity from 16 to the 15 allowed. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)
1. ![MINOR][MINOR] [AbstractProgressStep.java#L143](https://github.com/TGAC/miso-lims/blob/91d7ed1a5a0bbcc3e7e45c662e9fba927873c0c2/core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/workflow/AbstractProgressStep.java#L143): Replace this if-then-else statement by a single return statement. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1126)
1. ![MINOR][MINOR] [IntegerProgressStep.java#L39](https://github.com/TGAC/miso-lims/blob/91d7ed1a5a0bbcc3e7e45c662e9fba927873c0c2/core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/workflow/impl/IntegerProgressStep.java#L39): Replace this if-then-else statement by a single return statement. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1126)

[BLOCKER]: https://sonarsource.github.io/sonar-github/severity-blocker.png 'Severity: BLOCKER'
[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/rest/WorkflowRestController.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    Workflow workflow = workflowManager.loadWorkflow(id);
+    return new WorkflowStateDto(workflow, stepNumber);
+  }
+
+  @RequestMapping(value = "/continue", method = RequestMethod.GET)
+  public @ResponseBody WorkflowStateDto nextStep(@RequestParam("id") long id) throws IOException {
+    return new WorkflowStateDto(workflowManager.loadWorkflow(id));
+  }
+
+  @RequestMapping(value = "/cancelinput", method = RequestMethod.POST)
```
Comment (djcooke) [inline miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/rest/WorkflowRestController.java]: `DELETE /rest/workflow/{workflowId}/step/latest`

File: miso-web/src/main/webapp/scripts/workflow_display.js [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+      entryStepNumber === numEntries - 1 ? cancelInput(workflowId) : changeStep(workflowId, entryStepNumber)
+    };
+  }
+
+  function makeLog(logEntries, workflowId, currentStepNumber) {
+    var table = jQuery("<table class='workflowLogTable'>");
+
+    for (var i = 0; i < logEntries.length; i++) {
+      table.prepend(jQuery("<tr>").append(
+          [makeArrowCell(i, currentStepNumber), jQuery("<td>").text(logEntries[i]), jQuery("<td>"), makeModifyCell(i, logEntries.length)])
```
Comment (djcooke) [inline miso-web/src/main/webapp/scripts/workflow_display.js]: Is there a blank column in here between the log message and action button?

Comment (ndrein-oicr) [reply miso-web/src/main/webapp/scripts/workflow_display.js]: There is, good catch.  I put it there to balance the "arrow" column that precedes the log message.  Both the "arrow" column and the "blank" column will have the same width, as dictated by CSS, but I'm open to suggestions.

**Comment by sqwci** [comment]
SonarQube analysis reported 5 issues
* ![CRITICAL][CRITICAL] 3 critical
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

#### 4 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [LoadSequencerWorkflow.java#L43](https://github.com/TGAC/miso-lims/blob/7a2d4ac201fbc3efd160a775158373e40d3741f0/core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/workflow/impl/LoadSequencerWorkflow.java#L43): Define a constant instead of duplicating this literal "Not implemented yet" 4 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [ProgressImpl.java#L136](https://github.com/TGAC/miso-lims/blob/7a2d4ac201fbc3efd160a775158373e40d3741f0/core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/workflow/impl/ProgressImpl.java#L136): Refactor this method to reduce its Cognitive Complexity from 16 to the 15 allowed. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)
1. ![MINOR][MINOR] [AbstractProgressStep.java#L143](https://github.com/TGAC/miso-lims/blob/7a2d4ac201fbc3efd160a775158373e40d3741f0/core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/workflow/AbstractProgressStep.java#L143): Replace this if-then-else statement by a single return statement. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1126)
1. ![MINOR][MINOR] [IntegerProgressStep.java#L38](https://github.com/TGAC/miso-lims/blob/7a2d4ac201fbc3efd160a775158373e40d3741f0/core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/workflow/impl/IntegerProgressStep.java#L38): Replace this if-then-else statement by a single return statement. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1126)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by sqwci** [comment]
SonarQube analysis reported 5 issues
* ![CRITICAL][CRITICAL] 3 critical
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

#### 4 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [LoadSequencerWorkflow.java#L43](https://github.com/TGAC/miso-lims/blob/853227218dc477a0ced8de3c2c5278ef21e73cc7/core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/workflow/impl/LoadSequencerWorkflow.java#L43): Define a constant instead of duplicating this literal "Not implemented yet" 4 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [ProgressImpl.java#L136](https://github.com/TGAC/miso-lims/blob/853227218dc477a0ced8de3c2c5278ef21e73cc7/core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/workflow/impl/ProgressImpl.java#L136): Refactor this method to reduce its Cognitive Complexity from 16 to the 15 allowed. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)
1. ![MINOR][MINOR] [AbstractProgressStep.java#L143](https://github.com/TGAC/miso-lims/blob/853227218dc477a0ced8de3c2c5278ef21e73cc7/core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/workflow/AbstractProgressStep.java#L143): Replace this if-then-else statement by a single return statement. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1126)
1. ![MINOR][MINOR] [IntegerProgressStep.java#L38](https://github.com/TGAC/miso-lims/blob/853227218dc477a0ced8de3c2c5278ef21e73cc7/core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/workflow/impl/IntegerProgressStep.java#L38): Replace this if-then-else statement by a single return statement. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1126)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by sqwci** [comment]
SonarQube analysis reported 5 issues
* ![CRITICAL][CRITICAL] 3 critical
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

#### 4 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [LoadSequencerWorkflow.java#L43](https://github.com/TGAC/miso-lims/blob/92724c0ee95d93bcacb86f9ac904597bdc297144/core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/workflow/impl/LoadSequencerWorkflow.java#L43): Define a constant instead of duplicating this literal "Not implemented yet" 4 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [ProgressImpl.java#L136](https://github.com/TGAC/miso-lims/blob/92724c0ee95d93bcacb86f9ac904597bdc297144/core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/workflow/impl/ProgressImpl.java#L136): Refactor this method to reduce its Cognitive Complexity from 16 to the 15 allowed. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)
1. ![MINOR][MINOR] [AbstractProgressStep.java#L143](https://github.com/TGAC/miso-lims/blob/92724c0ee95d93bcacb86f9ac904597bdc297144/core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/workflow/AbstractProgressStep.java#L143): Replace this if-then-else statement by a single return statement. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1126)
1. ![MINOR][MINOR] [IntegerProgressStep.java#L38](https://github.com/TGAC/miso-lims/blob/92724c0ee95d93bcacb86f9ac904597bdc297144/core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/workflow/impl/IntegerProgressStep.java#L38): Replace this if-then-else statement by a single return statement. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1126)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: miso-web/src/main/webapp/scripts/workflow_display.js [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+
+  function showError(xhr) {
+    loadingTag.remove();
+    display.children().show();
+    display.children("input").focus();
+    errorDiv.text(parseError(xhr));
+  }
+
+  function parseError(xhr) {
+    var responseObj = JSON.parse(xhr.responseText);
```
Comment (apmasell) [inline miso-web/src/main/webapp/scripts/workflow_display.js]: This should get wrapped in a `try` block. The response might be HTML junk if the session has closed.

File: miso-web/src/main/webapp/styles/images/redo.svg [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+<!-- Generator: Adobe Illustrator 16.0.0, SVG Export Plug-In . SVG Version: 6.00 Build 0)  -->
+<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">
+<svg version="1.1" id="Capa_1" xmlns="http://www.w3.org/2000/svg"  x="0px" y="0px"
+	 width="485.212px" height="485.212px" viewBox="0 0 485.212 485.212" style="enable-background:new 0 0 485.212 485.212;"
+	 xml:space="preserve">
+<g>
+	<path d="M242.607,424.559c-75.252,0-136.468-61.209-136.468-136.465c0-75.252,61.216-136.466,136.468-136.466v90.978
+		l151.629-121.302L242.607,0v90.978c-108.687,0-197.117,88.432-197.117,197.117c0,108.691,88.43,197.118,197.117,197.118
+		c108.687,0,197.114-88.427,197.114-197.118h-60.645C379.077,363.35,317.859,424.559,242.607,424.559z"/>
+</g>
```
Comment (apmasell) [inline miso-web/src/main/webapp/styles/images/redo.svg]: Please delete all these extra `<g></g>` pairs.

**Comment by sqwci** [comment]
SonarQube analysis reported 5 issues
* ![CRITICAL][CRITICAL] 3 critical
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

#### 4 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [LoadSequencerWorkflow.java#L43](https://github.com/TGAC/miso-lims/blob/e50c4d1169404c5d8843213119dd3bf5fce02e48/core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/workflow/impl/LoadSequencerWorkflow.java#L43): Define a constant instead of duplicating this literal "Not implemented yet" 4 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [ProgressImpl.java#L136](https://github.com/TGAC/miso-lims/blob/e50c4d1169404c5d8843213119dd3bf5fce02e48/core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/workflow/impl/ProgressImpl.java#L136): Refactor this method to reduce its Cognitive Complexity from 16 to the 15 allowed. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)
1. ![MINOR][MINOR] [AbstractProgressStep.java#L143](https://github.com/TGAC/miso-lims/blob/e50c4d1169404c5d8843213119dd3bf5fce02e48/core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/workflow/AbstractProgressStep.java#L143): Replace this if-then-else statement by a single return statement. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1126)
1. ![MINOR][MINOR] [IntegerProgressStep.java#L38](https://github.com/TGAC/miso-lims/blob/e50c4d1169404c5d8843213119dd3bf5fce02e48/core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/workflow/impl/IntegerProgressStep.java#L38): Replace this if-then-else statement by a single return statement. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1126)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**